sdk/tables/azure-data-tables/inc/azure/data/tables/tables_sas_builder.hpp

```diff
@@ -0,0 +1,189 @@
+     */
+    All = ~0
+  };
+
+  /** @brief Bitwise OR of two values*/
+  inline TablesSasPermissions operator|(TablesSasPermissions lhs, TablesSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<TablesSasPermissions>;
+    return static_cast<TablesSasPermissions>(static_cast<type>(lhs) | static_cast<type>(rhs));
+  }
+
+  /** @brief Bitwise AND of two values*/
+  inline TablesSasPermissions operator&(TablesSasPermissions lhs, TablesSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<TablesSasPermissions>;
+    return static_cast<TablesSasPermissions>(static_cast<type>(lhs) & static_cast<type>(rhs));
+  }
+
+  /**
+   * @brief TableSasBuilder is used to generate a Shared Access Signature (SAS) for an Azure
+   * Storage Tables.
+   */
+  struct TablesSasBuilder final
+  {
+    /**
+     * @brief The optional signed protocol field specifies the protocol permitted for a
+     * request made with the SAS.
+     */
+    SasProtocol Protocol;
+
+    /**
+     * @brief Optionally specify the time at which the shared access signature becomes
+     * valid. This timestamp will be truncated to second.
+     */
+    Azure::Nullable<Azure::DateTime> StartsOn;
+
+    /**
+     * @brief The time at which the shared access signature becomes invalid. This field must
+     * be omitted if it has been specified in an associated stored access policy. This timestamp
+     * will be truncated to second.
+     */
+    Azure::DateTime ExpiresOn;
+
+    /**
+     * @brief Specifies an IP address or a range of IP addresses from which to accept
+     * requests. If the IP address from which the request originates does not match the IP address
+     * or address range specified on the SAS token, the request is not authenticated. When
+     * specifying a range of IP addresses, note that the range is inclusive.
+     */
+    Azure::Nullable<std::string> IPRange;
+
+    /**
+     * @brief An optional unique value up to 64 characters in length that correlates to an
+     * access policy specified for the table.
+     */
+    std::string Identifier;
+
+    /**
+     * @brief The name of the table being made accessible.
+     */
+    std::string TableName;
+
+    /**
+     * @brief The optional start of the partition key values range being made available.
+     */
+    std::string PartitionKeyStart;
+
+    /**
+     * @brief The optional end of the partition key values range being made available.
+     */
+    std::string PartitionKeyEnd;
+
+    /**
+     * @brief The optional start of the row key values range being made available.
+     */
+    std::string RowKeyStart;
+
+    /**
+     * @brief The optional end of the partition key values range being made available.
+     */
+    std::string RowKeyEnd;
+
+    /**
+     * @brief Sets the permissions for the table SAS.
+     *
+     * @param permissions The allowed permissions.
+     */
+    void SetPermissions(TablesSasPermissions permissions);

```

Comment:
For better or worse, the convention we've used for `struct` is that a `struct` is used only when it has no methods (including constructor/explicit destructors). 

I'm personally not a fan of this pattern (I'd rather see `struct` be used in the C++ meaning of "a class which has only public methods" rather than the C meaning of "something that looks like plain old data, even if it's not"), but this IS the pattern we've settled on as a team.

Bottom line: This should be a class, not a struct :(.

Reply:
done